pilot/pkg/xds/internalgen.go

```diff
@@ -142,14 +155,111 @@ func (sg *InternalGen) startPush(typeURL string, data []proto.Message) {
 // We can also expose ACKS.
 func (sg *InternalGen) Generate(proxy *model.Proxy, push *model.PushContext, w *model.WatchedResource, updates model.XdsUpdates) model.Resources {
 	res := []*any.Any{}
-	switch w.TypeUrl {
+
+	// Look for ?query parameters.  We don't use url.Parse() because we have no scheme
+	typeURL := w.TypeUrl
+	var qparams url.Values
+	if qIndex := strings.Index(w.TypeUrl, "?"); qIndex >= 0 {
+		typeURL = w.TypeUrl[:qIndex]
+		var err error
+		qparams, err = url.ParseQuery(w.TypeUrl[qIndex+1:])
+		if err != nil {
+			adsLog.Infof("Invalid TypeUrl query params: %q", w.TypeUrl)
+		}
+	}
+
+	switch typeURL {
 	case TypeURLConnections:
 		sg.Server.adsClientsMutex.RLock()
 		// Create a temp map to avoid locking the add/remove
 		for _, v := range sg.Server.adsClients {
 			res = append(res, util.MessageToAny(v.xdsNode))
 		}
 		sg.Server.adsClientsMutex.RUnlock()
+	case TypeDebugSyncronization:
+		res = sg.debugSyncz()
+	case TypeDebugConfigDump:
+		proxyID := qparams.Get("proxyID")
+		if proxyID == "" {
+			adsLog.Info("Config Dump w/o proxyID query parameter")
+			break
+		}
+		res = sg.debugConfigDump(proxyID)
+	default:
+		adsLog.Infof("Unknown TypeUrl: %q", w.TypeUrl)
+	}
+	return res
+}
+
+func (sg *InternalGen) debugSyncz() []*any.Any {
+	res := []*any.Any{}
+
+	sg.Server.adsClientsMutex.RLock()
+	for _, con := range sg.Server.adsClients {
+		con.mu.RLock()
+		// Skip "nodes" without metdata (they are probably istioctl queries!)
+		if con.node != nil && con.node.Metadata != nil && con.node.Metadata.ProxyConfig != nil {
+			xdsConfigs := []*status.PerXdsConfig{}
+			for stype, watchedResource := range con.node.Active {
+				pxc := &status.PerXdsConfig{
+					Status: debugSyncStatus(watchedResource),
+				}
+				switch stype {
+				case pilot_xds_v3.ListenerShortType:
+					pxc.PerXdsConfig = &status.PerXdsConfig_ListenerConfig{}
+				case pilot_xds_v3.RouteShortType:
+					pxc.PerXdsConfig = &status.PerXdsConfig_RouteConfig{}
+				case pilot_xds_v3.EndpointShortType:
+					pxc.PerXdsConfig = &status.PerXdsConfig_EndpointConfig{}
+				case pilot_xds_v3.ClusterShortType:
+					pxc.PerXdsConfig = &status.PerXdsConfig_ClusterConfig{}
+				default:
+					adsLog.Warnf("Unknown watchedResource type: %q", stype)
```

Comment:
Don't log here, its perfectly fine to have other watches. If you don't want to include them in istioctl just ignore them.

Reply:
I put this in to see which of many methods was "catching" `/debug/syncz` and left it here because nothing is logged when nothing catches.  I will take it out but it would be great if someone wrote code to log when no watches accept a typeURL.

Reply:
Fixed by removing this logging.